### PR TITLE
Charts: Refine helm installation

### LINKED ISSUE #283: refine helm installation
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like in detail**

current chart installation: https://github.com/kubernetes-sigs/azurefile-csi-driver/tree/master/charts

could refer to:  https://github.com/Azure/aad-pod-identity/tree/master/charts

```
helm repo add aad-pod-identity https://raw.githubusercontent.com/Azure/aad-pod-identity/master/charts
helm install aad-pod-identity aad-pod-identity/aad-pod-identity
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by animeshk08**
This commit updates helm installation
by using helm repos.

Signed-off-by: Animesh Kumar <animuz111@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test

> /kind flake

 /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #283 

**Special notes for your reviewer**:

**Release note**:
```
Charts: Refine helm installation
```

### CODE REVIEW

File: charts/index.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      - https://github.com/kubernetes-sigs/azurefile-csi-driver
+    urls:
+    - https://raw.githubusercontent.com/kubernetes-sigs/azurefile-csi-driver/master/charts/azurefile-csi-driver-v0.7.0.tgz
+    version: v0.7.0
+  - apiVersion: v1
+    appVersion: v0.6.0
+    created: "2020-05-25T02:02:09.483531499+05:30"
+    description: Azure File Container Storage Interface (CSI) Storage Plugin
+    digest: ad581478836e360fc6b8367791867e7184268857b4f50b322ef5245010768bd4
+    home: https://github.com/kubernetes-sigs/azurefile-csi-driver
```
Comment (animeshk08) [inline charts/index.yaml:21]: Hi @andyzhangx, I have not included maintainer details for now. 
```
maintainers:
    - email: <email>
      name: <username/name>
```

Let me know if you want me to include your details. Thanks :)

Comment (andyzhangx) [reply charts/index.yaml:21]: here is the maintainer details, thanks.
```
maintainers:
    - email: xiazhang@microsoft.com
      name: andyzhangx
```

**Comment by andyzhangx** [comment]
BTW, how can I install old release, e.g. in this PR, how can I install `v0.7.0` release, pls provide instruction in this PR, thanks.

**Comment by animeshk08** [comment]
Hi, @andyzhangx Let me know if the project structure looks good to you now. I have updated the maintainer details and also added in the documentation info about how to install a specific version of the charts.

I have a few comments/suggestion:
* The version of some of the charts(`0.2.0` and `0.3.0`) is not prepended by 'v', i.e. `v0.2.0` and `v0.3.0`. This results in inconsistency in the name of the package created. Let me know if you want me to update that.

* By default if the version is not specified the latest version present in `index.yaml` is installed. In this case, this will be `v0.7.0`. I understand this is not the development/latest version. Now, we can move in two paths to include the latest version.
1. Add separate instruction of installing development version and not include it in `index.yaml` e.g. https://github.com/elastic/helm-charts/tree/master/elasticsearch#installing

2. Create a package of the latest version as well and include it in `index.yaml`. This is followed in most Azure repos AFAIK, e.g. https://github.com/Azure/aad-pod-identity/tree/master/charts. If we follow this approach I suggest changing the version of the development charts from `latest` to `v0.8.0` or any other preferred version.

Let me know what are your views. I will update the PR accordingly. Thanks :)

**Comment by andyzhangx** [comment]
@animeshk08 we should only supports latest 3 (or less) versions in this PR, started from `v0.4.0`.
I think we could add a new parameter `--version` in helm, and if not specified, install `latest` version by default. So on master branch, it should always use `latest` version.

e.g. for aad-pod-identity:
```console
helm install aad-pod-identity/aad-pod-identity --name pod-identity --version 1.5.4
```
https://github.com/Azure/aad-pod-identity/blob/master/charts/index.yaml#L18

**Comment by animeshk08** [comment]
> I think we could add a new parameter --version in helm, and if not specified, install latest version by default. So on master branch, it should always use latest version.

I am sorry if I was not clear before. Such a parameter `--version` is already in helm and I have also mentioned its use-case in the docs, What I wanted to tell you was that if you look at the repo for r aad-pod-identity. (https://github.com/Azure/aad-pod-identity/tree/master/charts). Their latest charts version is 0.2.0.
https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity/Chart.yaml#L4

Now, they have created a package for the latest version as well.(https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity-2.0.0.tgz). and referenced the same in their `index.yaml`.(https://github.com/Azure/aad-pod-identity/blob/master/charts/index.yaml#L18).

Now if we follow the same, and I create a package for the latest version and include it the `index.yaml` for our charts, my only concern is that our package will look like this: `azurefile-csi-driver-latest.tgz`. and if I continue using the version `latest` in `index.yaml` as well. Running  commands like `helm search -l azurefile-csi-driver` we will get the packages as:
```
NAME                                            CHART VERSION   APP VERSION     DESCRIPTION 
azurefile-csi-driver/azurefile-csi-driver       latest          latest          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.6.0          v0.6.0          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.5.0          v0.5.0          Azure File Container Storage Interface (CSI) Storage Plugin
azurefile-csi-driver/azurefile-csi-driver       v0.4.0          v0.4.0          Azure File Container Storage Interface (CSI) Storage Plugin

```
As you can see the version is `latest` which does not tell much and also stands out. My suggestion was to drop using `latest` as the version instead provide it with a version number which in our case will be `v0.8.0`. This suggestion comes after seeing other helm repos. As you can see they have a specified chart version and not `latest`. Our charts are also going to be public and I think we should follow the same
```
NAME                    CHART VERSION   APP VERSION     DESCRIPTION                                                 
stable/jasperreports    7.0.11          7.2.0           DEPRECATED The JasperReports server can be used as a stan...
stable/artifactory      7.3.1           6.1.0           DEPRECATED Universal Repository Manager supporting all ma...
stable/artifactory-ha   0.4.1           6.2.0           DEPRECATED Universal Repository Manager supporting all ma...
stable/chartmuseum      2.12.2          0.12.0          Host your own Helm Chart Repository                         
stable/dmarc2logstash   1.3.0           1.0.3           Provides a POP3-polled DMARC XML report injector into Ela...
stable/pgadmin          1.2.3           4.18.0          DEPRECATED - moved to new repo, see source for new location 
stable/satisfy          1.0.0           3.0.4           Composer repo hosting with Satisfy                          
stable/sentry           4.2.1           9.1.2           Sentry is a cross-platform crash reporting and aggregatio...
stable/sonatype-nexus   1.23.1          3.20.1-01       DEPRECATED - Sonatype Nexus is an open source repository ...
azurefile-csi-driver/
azurefile-csi-driver    latest          latest          Azure File Container Storage Interface (CSI) Storage Plugin
```


If you don't want this we can stick with `latest` as the version only.

> we should only supports latest 3 (or less) versions in this PR, started from v0.4.0

I will update the PR to start from `v0.4.0` :)

File: charts/README.md [hunk 2]
```diff
@@ -6,12 +6,22 @@ Quick start instructions for the setup and configuration of azurefile CSI driver
...
+
+### Search for a different versions of charts available
+```console
+$ helm search -l azurefile-csi-driver/
+```
+
+### Install a specific version of chart
+Specify the version of the chart to be installed using the `--version` tag. 
+```console
+helm install --name azurefile-csi-driver azurefile-csi-driver/azurefile-csi-driver --namespace kube-system --version v0.6.0
```
Comment (animeshk08) [inline charts/README.md:24]: I have included the use of `--version` here.

**Comment by andyzhangx** [comment]
thanks, let's use `v0.8.0` instead of `latest`, also could you fix this spelling error:
```
Checking spelling...
error: deploy/example/snapshot/README.md:5:3: "Minumum" is a misspelling of "Minimum"
Found spelling errors!
```

**Comment by animeshk08** [comment]
Yes, I am on it.

**Comment by animeshk08** [comment]
@andyzhangx PR is ready for review.

File: charts/latest/azurefile-csi-driver/Chart.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 apiVersion: v1
-appVersion: latest
```
Comment (andyzhangx) [inline charts/latest/azurefile-csi-driver/Chart.yaml:2]: after second thought, I think we should keep `latest` tag, `latest` means current master branch version, and if we specify `v0.8.0`, we need to change this version num every time there is new release.

File: deploy/csi-azurefile-controller.yaml [hunk 4]
```diff
@@ -121,7 +121,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
Comment (andyzhangx) [inline deploy/csi-azurefile-controller.yaml:124]: keep `latest`

File: charts/README.md [hunk 5]
```diff
@@ -6,12 +6,22 @@ Quick start instructions for the setup and configuration of azurefile CSI driver
 
 1. [install Helm Client 3.0+ ](https://helm.sh/docs/intro/quickstart/#install-helm)
 
-## Install AzureFile via `helm install`
+## Install a released version of Azurefile using Helm repository
```
Comment (andyzhangx) [inline charts/README.md:9]: `Install CSI Driver master version using Helm repository`

File: charts/latest/azurefile-csi-driver/Chart.yaml (hunk 3 above)
Comment (andyzhangx) [reply charts/latest/azurefile-csi-driver/Chart.yaml:2]: sorry for the turn around, while I think using `latest` should be more appropriate for maintenance in the future.

Comment (animeshk08) [reply charts/latest/azurefile-csi-driver/Chart.yaml:2]: Hi, @andyzhangx so we have a few other problems while using `latest` as the version.
Helm charts specify that the chart version has to be of SemVer2 format.
https://helm.sh/docs/topics/charts/#charts-and-versioning

This is needed to parse the version while using `--version` tag with `helm install`. Now the keyword `latest` does not follow this rule. Hence I am getting the following error:
```
Error: chart "azurefile-csi-driver" matching version "latest" not found in azurefile-csi-driver index. (try 'helm repo update'). improper constraint: latest
```
 When I try running
 ```
helm install --name azurefile-csi-driver azurefile-csi-driver/azurefile-csi-driver --namespace kube-system --version latest --debug
```
I confirmed this problem by checking these issues:
* https://github.com/helm/helm/issues/4109
* https://github.com/helm/helm/issues/3600

AFAIK there is also no way to specify a chart as default in `index.yaml`. What helm does is sort the chart versions and install the latest one if the `version` is not explicitly mentioned. 
https://github.com/helm/helm/blob/15d9a6190f5e3ed01b5f2f670be48e7a0290518d/pkg/repo/index.go#L152

Let me know how you would like me to move forward from here.

Comment (andyzhangx) [reply charts/latest/azurefile-csi-driver/Chart.yaml:2]: hi, what about remove `latest` in helm installation, so the biggest version in helm is `v0.7.0` , and keep charts\latest unchanged
CI test would run on chart `latest` , and don't change deploy version, keep it as `latest` ?

File: charts/latest/azurefile-csi-driver/Chart.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
...
+appVersion: v0.8.0
 description: Azure File Container Storage Interface (CSI) Storage Plugin
 name: azurefile-csi-driver
-version: latest
```
Comment (andyzhangx) [inline charts/latest/azurefile-csi-driver/Chart.yaml:5]: undo this change

File: charts/latest/azurefile-csi-driver/values.yaml [hunk 7]
```diff
@@ -1,7 +1,7 @@
 image:
   azurefile:
     repository: mcr.microsoft.com/k8s/csi/azurefile-csi
-    tag: latest
```
Comment (andyzhangx) [inline charts/latest/azurefile-csi-driver/values.yaml:4]: undo this change

File: deploy/csi-azurefile-controller.yaml (hunk 4 above)
Comment (andyzhangx) [reply deploy/csi-azurefile-controller.yaml:124]: undo this change

File: deploy/csi-azurefile-node-windows.yaml [hunk 8]
```diff
@@ -66,7 +66,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
Comment (andyzhangx) [inline deploy/csi-azurefile-node-windows.yaml:69]: undo this change

File: deploy/csi-azurefile-node.yaml [hunk 9]
```diff
@@ -66,7 +66,7 @@ spec:
               cpu: 10m
               memory: 20Mi
         - name: azurefile
-          image: mcr.microsoft.com/k8s/csi/azurefile-csi:latest
```
Comment (andyzhangx) [inline deploy/csi-azurefile-node.yaml:69]: undo this change

File: charts/latest/azurefile-csi-driver/Chart.yaml (hunk 3 above)
Comment (animeshk08) [reply charts/latest/azurefile-csi-driver/Chart.yaml:2]: Hi, so you don't want me to include `latest` inside `index.yaml`? In this case, the default installation will be `v0.7.0`.  If that's the case. I would again suggest following what elastic does. https://github.com/elastic/helm-charts/tree/master/elasticsearch#installing. Include stable version installation via helm repo and development(latest) version installation directly.

**Comment by animeshk08** [comment]
@andyzhangx PR is ready for review :)
